src/main/kotlin/three/consulting/epoc/service/FirebaseService.kt

```diff
@@ -0,0 +1,63 @@
+package three.consulting.epoc.service
+
+import com.google.auth.oauth2.GoogleCredentials
+import com.google.firebase.FirebaseApp
+import com.google.firebase.FirebaseOptions
+import com.google.firebase.auth.FirebaseAuth
+import mu.KotlinLogging
+import org.springframework.stereotype.Service
+import three.consulting.epoc.common.Role
+import three.consulting.epoc.dto.EmployeeDTO
+import three.consulting.epoc.dto.FirebaseUserDTO
+import three.consulting.epoc.entity.Employee
+import three.consulting.epoc.repository.EmployeeRepository
+
+private val logger = KotlinLogging.logger {}
+
+@Service
+class FirebaseService(
+    private val employeeRepository: EmployeeRepository
+) {
+    fun firebaseAuth(): FirebaseAuth? {
+        try {
+            val options = FirebaseOptions.builder()
+                .setCredentials(GoogleCredentials.getApplicationDefault())
+                .build()
+            val firebaseApp = FirebaseApp.initializeApp(options)
+            return FirebaseAuth.getInstance(firebaseApp)
+        } catch (e: java.lang.Exception) {
+            logger.error(e) { "Could not set up Firebase" }
+        }
+        return null
+    }
+
+    private val firebaseAuth = firebaseAuth()
+
+    fun syncFirebaseUser(firebaseUid: String, firebaseEmail: String, customClaims: Map<String, Any>?): EmployeeDTO {
+
+        val role = if (customClaims === null) Role.USER else Role.valueOf(customClaims["role"] as String)
+
+        logger.info { "Syncing user: $firebaseEmail" }
+        val employee = employeeRepository.findByEmail(firebaseEmail)
+
+        return if (employee == null) {
+            val newEmployee = Employee(firebaseEmail, firebaseUid, role)
+            firebaseAuth!!.setCustomUserClaims(firebaseUid, customClaims)
+            EmployeeDTO(employeeRepository.save(newEmployee))
+        } else {
+            employee.firebaseUid = firebaseUid
+            firebaseAuth!!.setCustomUserClaims(firebaseUid, customClaims)
+            EmployeeDTO(employeeRepository.save(employee))
+        }
+    }
+
+    fun updateFirebaseUserRole(firebaseUserDTO: FirebaseUserDTO) {
+        val customClaims = mapOf("role" to firebaseUserDTO.role)
+        firebaseAuth!!.setCustomUserClaims(firebaseUserDTO.firebaseUid, customClaims)
```

Comment:
Also if you define firebaseAuth as a bean which either returns a valid FirebaseAuth or throws an exception, you can get rid of this dangerous double bang operator